sdk/src/main/java/io/opentelemetry/sdk/tags/TagMapSdk.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.tags;
+
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.tags.Tag;
+import io.opentelemetry.tags.TagKey;
+import io.opentelemetry.tags.TagMap;
+import io.opentelemetry.tags.TagMetadata;
+import io.opentelemetry.tags.TagValue;
+import io.opentelemetry.tags.unsafe.ContextUtils;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+public class TagMapSdk implements TagMap {
+
+  // The types of the TagKey and Tag must match for each entry.
+  private final Map<TagKey, Tag> tags;
+  @Nullable private final TagMap parent;
+
+  /**
+   * Creates a new {@link TagMapSdk} with the given tags.
+   *
+   * @param tags the initial tags for this {@code TagMapSdk}.
+   * @param parent providing a default set of tags
+   */
+  private TagMapSdk(Map<? extends TagKey, ? extends Tag> tags, TagMap parent) {
+    this.tags = Collections.unmodifiableMap(new HashMap<>(checkNotNull(tags, "tags")));
+    this.parent = parent;
+  }
+
+  @Override
+  public Iterator<Tag> getIterator() {
```

Comment:
Should we return Collection here? This will change the API :)

Seems to me more natural.

Reply:
Yeah, I think that would work better.  I never liked that aspect of the OpenTracing API.

Reply:
@bogdandrutu should this be changed in this PR or in a separate PR (perhaps after this is merged in)?

Reply:
I suggest we keep this, and revisit the `Collection` change in another PR ;)

Reply:
Let's do a different PR. We can do it before we merge this one.

Reply:
@bogdandrutu would you mind if we merge this then allow the follow on PR to change both?

Reply:
I think it is fine to do it later.

Reply:
Please just file an issue to not forget about this.

Reply:
Filled #401